### PR TITLE
Ehn/rtp ground scatter range

### PR DESCRIPTION
**Pull request by francistee26**
# Scope 
This pull request enables pydarn to use the ground scatter mapped range coordinate system when plotting rtp plots. This is the coordinate system described in https://doi.org/10.1029/93JA01470 on page 324.

## Approval

**Number of approvals:** 2

## Test
```
import pydarnio
import pydarn
from matplotlib import pyplot as plt
from pydarn import Coords

fitacf_file = 'data_vt_sps/20170101.0000.02.sps.a.fitacf'
fitacf_reader = pydarnio.SDarnRead(fitacf_file)
fitacf_data = fitacf_reader.read_fitacf()
pydarn.RTP.plot_range_time(fitacf_data, beam_num=7, parameter='p_l', zmax=50, zmin=0, date_fmt='%H%M',
                 colorbar_label='Power (dB)', cmap='viridis', coord = Coords.GROUND_SCATTER_MAPPED_RANGE, hgt = 250)
plt.savefig('test1.png')
```

### CODE REVIEW

File: pydarn/plotting/rtp.py [hunk 1]
```diff
@@ -335,11 +340,21 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
...
             y = gate2slant(cls.dmap_data[0], y_max, rxrise=rxrise)
+        elif coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
+            rxrise = SuperDARNRadars.radars[cls.dmap_data[0]['stid']]\
+                                    .hardware_info.rx_rise_time
+            y = gate2slant(cls.dmap_data[0], y_max, rxrise=rxrise)
+            Re = 6371
+            y = Re*np.arcsin(np.sqrt((y**2/4)-(hgt**2))/Re)  
+            y0inx = np.min(np.where(np.isfinite(y))[0])
+            y = y[y0inx:]
+            z = z[:,y0inx:]
```
Comment (carleyjmartin) [inline pydarn/plotting/rtp.py:357]: I'd maybe consider a comment around here to reference the equation in Bills paper. And a comment for line 355 just explaining quickly what it is doing, it's beautifully compact, but not immediately obvious.

Would this also benefit from being inside the `utils/conversions.py` module so that it could be used elsewhere if needed? This will also fix the issue of not being able to set a `hgt` for the `rtps` in summary plots.

**Review by carleyjmartin — CHANGES_REQUESTED** [review]
Hello! Read nothing into the speediness of my review! I had an hour or so free and was interested!
Also, congratulations on your first github PR, apparently!

I test by making a new venv and `pip3 install git+https://github.com/superdarn/pydarn@EHN/rtp_ground_scatter_range` and then just running the corresponding files.
So testing using the code below, commenting in and out the commented lines, on mac: 
```python
import matplotlib.pyplot as plt 
import pydarn

#Load in dmap file
file = "/Users/carley/Documents/data/20150305.0601.00.inv.fitacf"
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

#a = pydarn.RTP.plot_range_time(fitacf_data, beam_num=5, coord=pydarn.Coords.GROUND_SCATTER_MAPPED_RANGE, hgt=250)
#a = pydarn.RTP.plot_range_time(fitacf_data, beam_num=5, coord=pydarn.Coords.SLANT_RANGE)
a = pydarn.RTP.plot_summary(fitacf_data, beam_num=2, coord=pydarn.Coords.GROUND_SCATTER_MAPPED_RANGE)
plt.show()
```

The main issue is that it's not working for summary plotting, in the current set up summary plots are built by calling the time-series and range-time plot methods, and I think there must be an if statement in the summary plot method for the y-axis (but it is outside the scope of the PR changes so I can't comment on it) which means that it's not plotting summary plots correctly if given the `Coords.GROUND_SCATTER_MAPPING_RANGE`. You also can't read in a `hgt` value for summary plots, but given there's a default to 250 km it doesn't break it further. 

The other issue is minor and is just an aesthetic for the y-axis ticks, I've left a comment in line somewhere for that, other than that it appears to work great for RTPs. 

The last thing is including this option in documentation, which we can sort out in another Issue/PR at some point. If you have questions or want help with anything, just ask :D 

And just for future reference to me: checked new eqns against literature and appears to give reasonable values in testing.

File: pydarn/plotting/rtp.py [hunk 2]
```diff
@@ -417,7 +432,7 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
         else:
```
[the remark this answers is not in the document]
Comment (francistee26) [reply pydarn/plotting/rtp.py:432]: Hello @carleyjmartin thanks for your comments I have made some corrections based on your suggestions. Please let me know if you find anything else that I can improve on.

File: pydarn/plotting/rtp.py [hunk 3]
```diff
@@ -417,7 +435,7 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
...
             tick_interval = 1
         ax.xaxis.set_minor_locator(dates.MinuteLocator(interval=tick_interval))
-        if coord is Coord.SLANT_RANGE:
+        if coord is Coords.SLANT_RANGE:
             ax.yaxis.set_minor_locator(ticker.AutoMinorLocator(2))
         else:
             ax.yaxis.set_minor_locator(ticker.MultipleLocator(5))
```
Comment (carleyjmartin) [inline pydarn/plotting/rtp.py:441]: Oops sorry, I gave you the wrong line to look at! It was the minor ticks that had the issue. This should fix it, and puts one minor tick between major ticks. 

```suggestion
        if coord is Coords.SLANT_RANGE:
            ax.yaxis.set_minor_locator(ticker.AutoMinorLocator(2))
        if coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
            ax.yaxis.set_minor_locator(ticker.AutoMinorLocator(2))
        else:
            ax.yaxis.set_minor_locator(ticker.MultipleLocator(5))
```

**Review by carleyjmartin — COMMENTED** [review]
Looks great for summary plots! Sorry, I confused the major and minor ticks, the suggestion should fix the axis.

**Comment by francistee26** [comment]
Thanks @carleyjmartin I have made changes based off your suggestions.

File: pydarn/__init__.py [hunk 4]
```diff
@@ -27,6 +27,7 @@
 # importing utils
 from .utils.conversions import dmap2dict
 from .utils.conversions import gate2slant
+from .utils.conversions import gateGroundScatter
```
Comment (mts299) [inline pydarn/__init__.py:30]:
``` suggestion
gate2GroundScatter
```
for consistency sake

File: pydarn/plotting/rtp.py [hunk 5]
```diff
@@ -14,10 +14,10 @@
 from matplotlib import dates, colors, cm, ticker
 from typing import List
 
-from pydarn import (gate2slant, check_data_type, time2datetime,
+from pydarn import (gateGroundScatter, gate2slant, check_data_type, time2datetime,
```
Comment (mts299) [inline pydarn/plotting/rtp.py:17]:
```  suggestion
from pydarn import (gate2GroundScatter, gate2slant, check_data_type, time2datetime,
```

File: pydarn/plotting/rtp.py [hunk 6]
```diff
@@ -56,7 +56,8 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
                         start_time: datetime = None, end_time: datetime = None,
                         colorbar: plt.colorbar = None, ymin: int = None,
                         ymax: int = None, yspacing: int = 200,
-                        coord: object = Coord.SLANT_RANGE, colorbar_label: str = '',
+                        coord: object = Coords.SLANT_RANGE, hgt = 250,
```
Comment (mts299) [inline pydarn/plotting/rtp.py:59]:
``` suggestion
                        coord: object = Coords.SLANT_RANGE, reflection_height = 250, 
```
Strongly advise not using abbreviations if the word is simple enough, try to make it descriptive following PEP 8 standards.

File: pydarn/plotting/rtp.py [hunk 7]
```diff
@@ -109,9 +110,13 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
         yspacing: int
             sets the spacing between ticks
             Default: 200
-        coord: Coord
+        coord: Coords
             set the y-axis to a desired coordinate system
-            Default: Coord.SLANT_RANGE
+            Default: Coords.SLANT_RANGE
+        hgt: int
```
Comment (mts299) [inline pydarn/plotting/rtp.py:116]:
``` suggestion
        reflection_height: int
```

File: pydarn/plotting/rtp.py [hunk 8]
```diff
@@ -335,11 +340,21 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
...
+        if coord is Coords.SLANT_RANGE:
             # Get rxrise from hardware files (consistent with RST)
             rxrise = SuperDARNRadars.radars[cls.dmap_data[0]['stid']]\
                                     .hardware_info.rx_rise_time
             y = gate2slant(cls.dmap_data[0], y_max, rxrise=rxrise)
+        elif coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
+            rxrise = SuperDARNRadars.radars[cls.dmap_data[0]['stid']]\
+                                    .hardware_info.rx_rise_time
+            y = gate2slant(cls.dmap_data[0], y_max, rxrise=rxrise)
+            Re = 6371
```
Comment (mts299) [inline pydarn/plotting/rtp.py:353]: Let me know to make a const file in another PR since I need this for my other PR and then we have consistency on constants.

File: pydarn/plotting/rtp.py [hunk 9]
```diff
@@ -335,11 +340,21 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
...
+            y = gateGroundScatter(y, Re, hgt)
+            y0inx = np.min(np.where(np.isfinite(y))[0])
```
Comment (mts299) [inline pydarn/plotting/rtp.py:355]: Either comment or a better variable name. I assume zeroth y index? Which I am not sure what this is used for. 
or is it minimum y value?

File: pydarn/plotting/rtp.py [hunk 10]
```diff
@@ -393,14 +408,17 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
         ax.set_xlim([rounded_down_start_time, x[-1]])
         ax.xaxis.set_major_formatter(dates.DateFormatter(date_fmt))
         if ymax is None:
-            ymax = max(y)
+            ymax = np.max(y)
```
Comment (mts299) [inline pydarn/plotting/rtp.py:411]: Any reason to change this to using numpy?

File: pydarn/plotting/rtp.py [hunk 11]
```diff
@@ -393,14 +408,17 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
...
-            ymin = min(y)
+            ymin = np.min(y)
 
         ax.set_ylim(ymin, ymax)
 
-        if coord is Coord.SLANT_RANGE:
+        if coord is Coords.SLANT_RANGE:
+            ax.yaxis.set_ticks(np.arange(np.ceil(ymin/100.0)*100,
+                                         ymax+1, yspacing))
+        elif coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
```
Comment (mts299) [inline pydarn/plotting/rtp.py:421]:
``` suggestion
        if coord is Coords.SLANT_RANGE or coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
```

they plot the same thing assuming the ticks are good.

File: pydarn/plotting/rtp.py [hunk 12]
```diff
@@ -679,7 +699,7 @@ def plot_time_series(cls, dmap_data: List[dict],
     @classmethod
     def plot_summary(cls, dmap_data: List[dict], beam_num: int = 0,
                      groundscatter: bool = True, channel: int = 'all',
-                     coord: object = Coord.SLANT_RANGE, figsize: tuple = (11, 8.5),
+                     coord: object = Coords.SLANT_RANGE, hgt: float = 250.0, figsize: tuple = (11, 8.5),
```
Comment (mts299) [inline pydarn/plotting/rtp.py:702]:
```  suggestion
                     coord: object = Coords.SLANT_RANGE, reflection_height: float = 250.0, figsize: tuple = (11, 8.5),
```

Any reason it is 250.0 here but at range time it's 250? Actually, you have it as an int there?

File: pydarn/plotting/rtp.py (hunk 7 above)
Comment (mts299) [reply pydarn/plotting/rtp.py:116]: So later on in summary plots, you have this value as a  float? Should this be a float?

File: pydarn/plotting/rtp.py [hunk 13]
```diff
@@ -714,6 +734,10 @@ def plot_summary(cls, dmap_data: List[dict], beam_num: int = 0,
         coord: Coord
             set the y-axis to a desired coordinate system
             Default: Coord.SLANT_RANGE
+        hgt: int
```
Comment (mts299) [inline pydarn/plotting/rtp.py:737]:
``` suggestion
        reflection_height: int
```
float or int?

File: pydarn/plotting/rtp.py [hunk 14]
```diff
@@ -965,8 +989,10 @@ def plot_summary(cls, dmap_data: List[dict], beam_num: int = 0,
             else:
                 # Current standard is to only have groundscatter
                 # on the velocity plot. This may change in the future.
-                if coord is Coord.SLANT_RANGE:
+                if coord is Coords.SLANT_RANGE:
                     ymax = 3517.5
+                elif coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
+                    ymax = 3517.5/2
```
Comment (mts299) [inline pydarn/plotting/rtp.py:995]: Why divided by  2?

File: pydarn/plotting/rtp.py [hunk 15]
```diff
@@ -986,7 +1012,8 @@ def plot_summary(cls, dmap_data: List[dict], beam_num: int = 0,
                                             ax=axes[i],
                                             groundscatter=grndflg,
                                             channel=channel,
-                                            coord=coord,
+                                            coord=coord, 
+                                            hgt = hgt,
```
Comment (mts299) [inline pydarn/plotting/rtp.py:1016]:
``` suggestion
                                            reflection_height = reflection_height,
```

File: pydarn/plotting/rtp.py [hunk 16]
```diff
@@ -335,11 +340,21 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
...
             # Get rxrise from hardware files (consistent with RST)
             rxrise = SuperDARNRadars.radars[cls.dmap_data[0]['stid']]\
                                     .hardware_info.rx_rise_time
             y = gate2slant(cls.dmap_data[0], y_max, rxrise=rxrise)
+        elif coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
+            rxrise = SuperDARNRadars.radars[cls.dmap_data[0]['stid']]\
+                                    .hardware_info.rx_rise_time
+            y = gate2slant(cls.dmap_data[0], y_max, rxrise=rxrise)
+            Re = 6371
+            y = gateGroundScatter(y, Re, hgt)
```
Comment (mts299) [inline pydarn/plotting/rtp.py:354]:
``` suggestion
            y = gate2GroundScatter(y, Re, reflection_height)
```

File: pydarn/utils/conversions.py [hunk 17]
```diff
@@ -81,7 +81,26 @@ def dmap2dict(dmap_records: List[dict]) -> List[dict]:
         dmap_list.append(OrderedDict(dmap_dict))
     return dmap_list
 
+def gateGroundScatter(y, Re, hgt=250):
```
Comment (mts299) [inline pydarn/utils/conversions.py:84]:
``` suggeestion
def gate2GroundScatter(y, Re, reflection_height=250):
```

File: pydarn/utils/conversions.py [hunk 18]
```diff
@@ -81,7 +81,26 @@ def dmap2dict(dmap_records: List[dict]) -> List[dict]:
...
+    """
+    Calculate the ground scatter mapped range (km) for each range gate for SuperDARN data
 
+    Parameters
+    ----------
+        Re: int
+            earth's radius
+        hgt: int
+            reflection height
```
Comment (mts299) [inline pydarn/utils/conversions.py:93]:
``` suggestion
        Re: float
            earth's radius
        hgt: float
            reflection height
```
I believe these are floats?

File: pydarn/utils/conversions.py [hunk 19]
```diff
@@ -128,3 +147,5 @@ def gate2slant(record, nrang, rxrise=0, center=False):
                               gate * sample_sep) * speed_of_light /\
                 distance_factor + range_offset
     return slant_ranges
+    
+
```
Comment (mts299) [inline pydarn/utils/conversions.py:151]:
``` suggestion
```

PEP 8

File: pydarn/utils/conversions.py [hunk 20]
```diff
@@ -81,7 +81,26 @@ def dmap2dict(dmap_records: List[dict]) -> List[dict]:
         dmap_list.append(OrderedDict(dmap_dict))
     return dmap_list
 
+def gateGroundScatter(y, Re, hgt=250):
+    """
+    Calculate the ground scatter mapped range (km) for each range gate for SuperDARN data
```
Comment (mts299) [inline pydarn/utils/conversions.py:86]: If there is a  published paper this work is based on then I would put it in for people's further understanding.

File: pydarn/utils/conversions.py (hunk 17 above)
Comment (mts299) [inline pydarn/utils/conversions.py:84]: At the top of this file add your copyright for this function please :)

File: pydarn/utils/coordinates.py [hunk 21]
```diff
@@ -20,15 +20,17 @@
 """
 import enum
 
-class Coord(enum.Enum):
+class Coords(enum.Enum):
```
Comment (mts299) [inline pydarn/utils/coordinates.py:23]: Aaaaa does this break the ACF plotting code? be careful replacing already established method names. This can break code.  I would almost suggest reverting for safety's sake if others agree.

**Review by mts299 — CHANGES_REQUESTED** [review]
Good job! lots of minor things and changing names to be consistent and descriptive. 

One thing to  look into  is the affect  of  changing pre-approved  Coord to Coords this may break  other existing  code

File: pydarn/utils/conversions.py (hunk 20 above)
Comment (carleyjmartin) [reply pydarn/utils/conversions.py:86]: Oo also amend to gate2GroundScatter given above comments

File: pydarn/plotting/rtp.py (hunk 11 above)
Comment (mts299) [reply pydarn/plotting/rtp.py:421]: @carleyjmartin is this the spot?

Comment (carleyjmartin) [reply pydarn/plotting/rtp.py:421]: Nah it's later in the minor ticks (L438-444), I'm trying to find the source now, but for reference the minor tick bug from before is now happening in slant range option.

Comment (carleyjmartin) [reply pydarn/plotting/rtp.py:421]: Found the issue! Will comment a suggestion!

File: pydarn/plotting/rtp.py [hunk 22]
```diff
@@ -417,7 +435,9 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
         else:
             tick_interval = 1
         ax.xaxis.set_minor_locator(dates.MinuteLocator(interval=tick_interval))
-        if coord is Coord.SLANT_RANGE:
+        if coord is Coords.SLANT_RANGE:
+            ax.yaxis.set_minor_locator(ticker.AutoMinorLocator(2))
+        if coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
```
Comment (carleyjmartin) [inline pydarn/plotting/rtp.py:440]: my fault, sorry!
```suggestion
        elif coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
```

Comment (carleyjmartin) [reply pydarn/plotting/rtp.py:440]: or add it to the if condition on the line above (like Marina's comment earlier to be consistent)

File: pydarn/plotting/rtp.py [hunk 23]
```diff
@@ -417,7 +435,9 @@ def plot_range_time(cls, dmap_data: List[dict], parameter: str = 'v',
...
             ax.yaxis.set_minor_locator(ticker.AutoMinorLocator(2))
         else:
             ax.yaxis.set_minor_locator(ticker.MultipleLocator(5))
```
Comment (carleyjmartin) [inline pydarn/plotting/rtp.py:443]: sorry for the notification, just amending to follow Marina's suggestion on the other one
```suggestion
        if coord is Coords.SLANT_RANGE or coord is Coords.GROUND_SCATTER_MAPPED_RANGE:
            ax.yaxis.set_minor_locator(ticker.AutoMinorLocator(2))
        else:
            ax.yaxis.set_minor_locator(ticker.MultipleLocator(5))
```

File: pydarn/plotting/rtp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (carleyjmartin) [reply pydarn/plotting/rtp.py:353]: @francistee26 a constants module has now been added to develop, you can read in Re from there now :)

File: pydarn/utils/coordinates.py (hunk 21 above)
Comment (francistee26) [reply pydarn/utils/coordinates.py:23]: I had a circle loop error when I used just Coord to call the coord type. However, that stopped when I changed the name to Coords. I was also thinking since we will not just be using one Coordinate it will be reasonable to name it Coords to implying not just one coordinate system.

File: pydarn/plotting/rtp.py (hunk 23 above)
Comment (carleyjmartin) [reply pydarn/plotting/rtp.py]: And in doing so I deleted the actual comment - this also fixes the bug that was created by me in the first place (sorry, see the if instead of elif on line 440 pre changes - that was causing the bug)

File: pydarn/utils/conversions.py (hunk 20 above)
Comment (francistee26) [reply pydarn/utils/conversions.py:86]: Sorry about that I added it as a comment I have now moved it into the the doc string

File: pydarn/utils/conversions.py [hunk 24]
```diff
@@ -83,7 +83,26 @@ def dmap2dict(dmap_records: List[dict]) -> List[dict]:
         dmap_list.append(OrderedDict(dmap_dict))
     return dmap_list
 
+def gate2GroundScatter(y, Re, reflection_height=250):
```
Comment (mts299) [inline pydarn/utils/conversions.py:86]: You don't want to pass in a constant, instead remove it as a parameter and then import it and use it in the file. Hopefully, that makes sense?

File: pydarn/utils/conversions.py (hunk 17 above)
Comment (francistee26) [reply pydarn/utils/conversions.py]: Hey @mts299 Do you have any example of what I should put in the copyright for the function?

**Comment by w2naf** [comment]
Yes, I'll be happy to.  Thanks!

**Comment by w2naf** [comment]
Hi @mts299,

I just merged the current current develop branch into this PR and resolved the conflicts. I fixed a couple of other bugs I found, too. Hopefully it is good to go!

**Review by mts299 — APPROVED** [review]
Looks good to me code wise @carleyjmartin can you test the code again with the new updates.

**Review by carleyjmartin — APPROVED** [review]
Tested using plot_range_time and plot_summary methods for all three variations of y-axis units. No issues found, works without inputting a reflection height for GSMR as well.
